server/app/controllers/applicant/UpsellController.java

```diff
@@ -165,4 +177,42 @@ public CompletionStage<Result> considerRegister(
               throw new RuntimeException(ex);
             });
   }
+
+  /** Download a PDF file of the application to the program. */
+  @Secure
+  public CompletionStage<Result> download(
+      Http.Request request, long programId, long applicationId, long applicantId)
+      throws ProgramNotFoundException {
+    ProgramDefinition program = programService.getProgramDefinition(programId);
+    Optional<CiviFormProfile> profileMaybe = profileUtils.currentUserProfile(request);
+    CiviFormProfile profile =
+        profileMaybe.orElseThrow(
+            () -> new NoSuchElementException("User authorized as applicant but no profile found"));
+    if (!profile.isTrustedIntermediary()) {
+      try {
+        super.checkApplicantAuthorization(request, applicantId).join();
+      } catch (CompletionException | SecurityException e) {
+        return CompletableFuture.completedFuture(unauthorized("Invalid credentials"));
+      }
+    } else {
+      Account currentAccount = profile.getAccount().join();
+      if (currentAccount.getManagedByGroup().isEmpty()) {
+        return CompletableFuture.completedFuture(unauthorized("Invalid credentials"));
+      }
+    }
+    Optional<Application> applicationMaybe =
+        applicationService.getApplication(applicationId, program);
```

Comment:
**suggestion (blocking):** there are several places in this code that are making still making synchronous database calls. Please rewrite using async calls to access the DB.

https://github.com/civiform/civiform/pull/5491#discussion_r1316393352